packages/react-client/src/hooks/internal/device/useDevice.ts

```diff
@@ -0,0 +1,172 @@
+  allDevicesList: MediaDeviceInfo[];
+  constraints?: MediaTrackConstraints | boolean;
+  saveUsedDevice: (device: MediaDeviceInfo) => void;
+};
+
+export type DeviceManager = {
+  startDevice: (deviceId?: string) => Promise<MediaStreamTrack | null>;
+  stopDevice: () => void;
+  activeDevice: DeviceItem | null;
+  deviceTrack: MediaStreamTrack | null;
+  deviceList: DeviceItem[];
+  deviceEnabled: boolean;
+  enableDevice: () => void;
+  disableDevice: () => void;
+  currentMiddleware: TrackMiddleware;
+  applyMiddleware: (middleware: TrackMiddleware) => MediaStreamTrack | null;
+  deviceError: DeviceError | null;
+};
+
+function getReplaceStreamAction(
+  newStream: MediaStream | null,
+  deviceType: "audio" | "video",
+): SetStateAction<MediaStream | null> {
+  return (oldStream: MediaStream | null) => {
+    if (oldStream) {
+      stopStream(oldStream, deviceType);
+    }
+    return newStream;
+  };
+}
+
+function getStopStreamAction(deviceType: "audio" | "video"): SetStateAction<MediaStream | null> {
+  return (oldStream: MediaStream | null) => {
+    if (oldStream) {
+      stopStream(oldStream, deviceType);
+    }
+    return null;
+  };
+}
+
+async function getDeviceStream(
+  type: "audio" | "video",
+  constraints: MediaTrackConstraints | boolean | undefined,
+  deviceId?: string,
+) {
+  if (constraints === false) {
+    console.warn("Attempted to enable camera, but its disabled by a constraint");
+    return null;
+  }
+  constraints = typeof constraints === "object" ? constraints : {};
+  if (deviceId) {
+    constraints.deviceId = { exact: deviceId };
+  }
+  const stream = await navigator.mediaDevices.getUserMedia({
+    [type]: constraints,
+  });
+  return stream;
+}
+
+export const useDeviceManager = ({
+  mediaStream,
+  getInitialStream,
+  deviceType,
+  allDevicesList,
+  setStream,
+  constraints,
+  saveUsedDevice,
+  deviceError,
+  setError,
+}: DeviceManagerProps): DeviceManager => {
+  const rawTrack = useMemo(() => mediaStream && getTrackFromStream(mediaStream, deviceType), [mediaStream, deviceType]);
+
+  const { processedTrack, applyMiddleware, currentMiddleware } = useTrackMiddleware(rawTrack);
+
+  const currentTrack = useMemo(() => processedTrack ?? rawTrack, [rawTrack, processedTrack]);
+
+  const deviceList = useMemo(
+    () => allDevicesList.filter(({ kind }) => kind === `${deviceType}input`),
+    [allDevicesList, deviceType],
+  );
+
+  const activeDevice = useMemo(() => {
+    const currentDevice =
+      mediaStream &&
+      deviceList.find(
+        (device) => device.deviceId === getTrackFromStream(mediaStream, deviceType)?.getSettings().deviceId,
+      );
+    if (!currentDevice) return null;
+    return { label: currentDevice.label, deviceId: currentDevice.deviceId };
+  }, [mediaStream, deviceList, deviceType]);
+
+  const [deviceEnabled, setDeviceEnabled] = useState(true);
+
+  const startDevice = useCallback(
+    async (deviceId?: string) => {
+      let stream = await getInitialStream();
```

Comment:
Does `stream` variable needs to be overwritten?
```suggestion
      const initialStream = await getInitialStream();
```